Namespace sometimes gets created in incorrect kube-context

When trying to run `reckoner plot` on a course file with an explicit `context` and `namespace`, which differs from my current kube-context, and also the namespace does not exist on the given context, the deployment fails and the namespace is created on my current kube-context.

Ok that was confusing, so here is an example :)

```yaml
# course.yaml
context: star-wars-cluster
namespace: skywalker
repositories:
  prometheus-community/helm-charts:
    git: https://github.com/prometheus-community/helm-charts.git
    path: charts
charts:
  kube-prometheus-stack:
    version: kube-prometheus-stack-18.0.2
    repository: prometheus-community/helm-charts
    namespace: skywalker
```

```
> kubectl config set-context pokemon-cluster --namespace=pikachu
> kubectl config use-context pokemon-cluster

> reckoner plot course.yaml
2021-09-03 11:15:25 maiko root[13238] INFO Found Helm Version 3.6.3
2021-09-03 11:15:45 maiko root[13238] INFO Running 'install' on kube-prometheus-stack in skywalker
2021-09-03 11:15:46 maiko root[13238] INFO Updating chart dependencies: /home/maiko/.cache/helm/repository/https_github_com_prometheus-community_helm-charts_git/charts/kube-prometheus-stack
2021-09-03 11:16:18 maiko root[13238] INFO Namespace skywalker not found. Creating it now.
2021-09-03 11:16:22 maiko root[13238] ERROR ERROR: install Failed on kube-prometheus-stack
⛵🔥 Encountered errors while running the course ⛵🔥

* * * * *

Chart Name: kube-prometheus-stack
Status: Failed
Error Reason: Command Failed with output below:
STDOUT: Release "kube-prometheus-stack" does not exist. Installing it now.

STDERR: Error: rendered manifests contain a resource that already exists. Unable to continue with install: PodSecurityPolicy "kube-prometheus-stack-grafana" in namespace "" exists and cannot be imported into the current release: invalid ownership metadata; annotation validation error: key "meta.helm.sh/release-namespace" must equal "skywalker": current value is "pikachu"

COMMAND: helm upgrade --install kube-prometheus-stack /home/maiko/.cache/helm/repository/https_github_com_prometheus-community_helm-charts_git/charts/kube-prometheus-stack --namespace skywalker --kube-context star-wars-cluster --version kube-prometheus-stack-18.0.2
Makefile:61: recipe for target 'deploy' failed
make: *** [deploy] Error 1
```

Now the namespace `skywalker` was created in the `pokemon-cluster` context:

```bash
> kubectl get namespaces --context=pokemon-cluster | grep skywalker
skywalker     Active   37s
> kubectl get namespaces --context=star-wars-cluster | grep skywalker
```

--- a/reckoner/chart.py
+++ b/reckoner/chart.py
@@ -163,12 +163,12 @@ class Chart(object):
 
     @property
     def namespace_management(self):
-        """ Namespace to install the course chart """
+        """ Parameters for augmenting namespaces we install charts into """
         return self._namespace_management
 
     @property
     def context(self):
-        """ Namespace to install the course chart """
+        """ Context to use to install the course chart """
         return self._context
 
     @property
@@ -227,7 +227,7 @@ class Chart(object):
         in the self.config option to avoid going back to the api for each chart.
         """
         if self.config.create_namespace and not self.dryrun:
-            nsm = NamespaceManager(self.namespace, self.namespace_management)
+            nsm = NamespaceManager(self.namespace, self.namespace_management, self.context)
             nsm.create_and_manage()
 
     def __pre_command(self, default_namespace=None, default_namespace_management={}, context=None) -> None:

--- a/reckoner/course.py
+++ b/reckoner/course.py
@@ -419,7 +419,6 @@ class Course(object):
 
         logging.debug("Reckoner Minimum Version is {}".format(reckoner_minimum_version))
         logging.debug("Reckoner Installed Version is {}".format(reckoner_version))
-
         r1 = semver.compare(reckoner_version, reckoner_minimum_version)
         if r1 < 0:
             raise MinimumVersionException("reckoner Minimum Version {} not met.".format(reckoner_minimum_version))

--- a/reckoner/kube.py
+++ b/reckoner/kube.py
@@ -22,7 +22,7 @@ from kubernetes import client, config
 
 class NamespaceManager(object):
 
-    def __init__(self, namespace_name, namespace_management) -> None:
+    def __init__(self, namespace_name, namespace_management, kube_context) -> None:
         """ Manages a namespace for the chart
         Accepts:
         - namespace: Which may be a string or a dictionary
@@ -36,6 +36,7 @@ class NamespaceManager(object):
             'overwrite',
             False
         )
+        self._kube_context = kube_context
         self.__load_config()
         self.config = Config()
 
@@ -46,7 +47,7 @@ class NamespaceManager(object):
 
     @property
     def namespace(self) -> str:
-        """ Namespace object we are managing 
+        """ Namespace object we are managing
         https://github.com/kubernetes-client/python/blob/master/kubernetes/docs/V1Namespace.md"""
         return self._namespace
 
@@ -62,10 +63,16 @@ class NamespaceManager(object):
         from the chart and course """
         return self._overwrite
 
+    @property
+    def kube_context(self) -> str:
+        """ The kubernetes context that should be used
+        to create and manage the namespace """
+        return self._kube_context
+
     def __load_config(self):
         """ Protected method to load kubernetes config"""
         try:
-            config.load_kube_config()
+            config.load_kube_config(context=self.kube_context)
             self.v1client = client.CoreV1Api()
         except Exception as e:
             logging.error('Unable to load kubernetes configuration')

--- a/reckoner/tests/namespace_manager_mock.py
+++ b/reckoner/tests/namespace_manager_mock.py
@@ -4,7 +4,7 @@ from kubernetes import client, config
 
 class NamespaceManagerMock(object):
 
-    def __init__(self, namespace_name='', namespace_management={}) -> None:
+    def __init__(self, namespace_name='', namespace_management={}, kube_context=None) -> None:
         """ Manages a namespace for the chart
         Accepts:
         - namespace: Which may be a string or a dictionary
@@ -18,6 +18,7 @@ class NamespaceManagerMock(object):
             'overwrite',
             False
         )
+        self._kube_context = kube_context
         self.__load_config()
 
     @property
@@ -27,7 +28,7 @@ class NamespaceManagerMock(object):
 
     @property
     def namespace(self) -> str:
-        """ Namespace object we are managing 
+        """ Namespace object we are managing
         https://github.com/kubernetes-client/python/blob/master/kubernetes/docs/V1Namespace.md"""
         return self._namespace
 
@@ -42,6 +43,12 @@ class NamespaceManagerMock(object):
         """ List of metadata settings parsed from the
         from the chart and course """
         return self._overwrite
+
+    @property
+    def kube_context(self) -> str:
+        """ The kubernetes context that should be used
+        to create and manage the namespace """
+        return self._kube_context
 
     def __load_config(self):
         """ Protected method do load kubernetes config"""

--- a/reckoner/tests/test_kube.py
+++ b/reckoner/tests/test_kube.py
@@ -44,7 +44,7 @@ class TestNamespaceManager(unittest.TestCase):
             "settings": {"overwrite": True}
         }
 
-        _ns_manager = NamespaceManager('test', settings)
+        _ns_manager = NamespaceManager('test', settings, None)
 
         self.assertTrue(_ns_manager.overwrite)
         self.assertEqual(_ns_manager.metadata, {
@@ -68,7 +68,7 @@ class TestNamespaceManager(unittest.TestCase):
             "settings": {"overwrite": False}
         }
 
-        _ns_manager = NamespaceManager('test', settings)
+        _ns_manager = NamespaceManager('test', settings, None)
 
         self.assertFalse(_ns_manager.overwrite)
         self.assertEqual(_ns_manager.metadata, {})
@@ -79,7 +79,7 @@ class TestNamespaceManager(unittest.TestCase):
             "metadata": {}
         }
 
-        _ns_manager = NamespaceManager('test', settings)
+        _ns_manager = NamespaceManager('test', settings, None)
 
         self.assertFalse(_ns_manager.overwrite)
         self.assertEqual(_ns_manager.metadata, {})

